src/clang_parser.cpp

```diff
@@ -584,46 +607,46 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     args.push_back("-D__CLANG_WORKAROUNDS_H");
   }
 
-  auto incomplete_types = get_incomplete_types(input, input_files, args);
-  bpftrace.btf_set_.insert(incomplete_types.cbegin(), incomplete_types.cend());
+  // The generated BTF header is initially empty
+  std::string btf_cdef;
+  input_files.emplace_back(CXUnsavedFile{
+      .Filename = "/bpftrace/include/__btf_generated_header.h",
+      .Contents = btf_cdef.c_str(),
+      .Length = btf_cdef.size(),
+  });
 
   ClangParserHandler handler;
-  CXErrorCode error;
-
-  if (process_btf)
+  bool check_additional_types = true;
+  while (check_additional_types && process_btf)
   {
-    auto btf_and_input = "#include <__btf_generated_header.h>\n" + input;
-    auto btf_and_input_files = getTranslationUnitFiles(
-        CXUnsavedFile{ .Filename = "definitions.h",
-                       .Contents = btf_and_input.c_str(),
-                       .Length = btf_and_input.size() });
-
-    auto btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
-    btf_and_input_files.emplace_back(CXUnsavedFile{
-        .Filename = "/bpftrace/include/__btf_generated_header.h",
-        .Contents = btf_cdef.c_str(),
-        .Length = btf_cdef.size(),
-    });
-
-    error = handler.parse_translation_unit(
-        "definitions.h",
-        args.data(),
-        args.size(),
-        btf_and_input_files.data(),
-        btf_and_input_files.size(),
-        CXTranslationUnit_DetailedPreprocessingRecord);
-  }
-  else
-  {
-    error = handler.parse_translation_unit(
-        "definitions.h",
-        args.data(),
-        args.size(),
-        input_files.data(),
-        input_files.size(),
-        CXTranslationUnit_DetailedPreprocessingRecord);
+    auto incomplete_types = get_incomplete_types(
+        input, input_files, args, bpftrace.btf_set_);
+    size_t types_cnt = bpftrace.btf_set_.size();
+    bpftrace.btf_set_.insert(incomplete_types.cbegin(),
+                             incomplete_types.cend());
+
+    // Update generated header with current BTF types
+    btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
+    input_files.back() = CXUnsavedFile{
+      .Filename = "/bpftrace/include/__btf_generated_header.h",
+      .Contents = btf_cdef.c_str(),
+      .Length = btf_cdef.size(),
+    };
+
+    // If additional BTF types were found, we need to repeat the process since
+    // that might have introduced some new unresolved typedefs.
+    check_additional_types = types_cnt != bpftrace.btf_set_.size();
```

Comment:
@viktormalik 
I found this PR causes a slowdown if an unknown struct is big. For example:

before this PR

```
% time sudo  ./src/bpftrace --btf -e 'struct f { struct task_struct x; } BEGIN{exit();}'
Attaching 1 probe..

0.94s user 0.20s system 80% cpu 1.420 tota
```

after
```
% time sudo ./src/bpftrace --btf -e 'struct f { struct task_struct x; } BEGIN{exit();}'
Attaching 1 probe...

7.76s user 0.59s system 96% cpu 8.646 total
```

The problem is this commented part, and for `task_struct`, this loop iterates 9 times. I wonder if we can reduce the iteration. Actually, I'm not sure when we need to iterate until the condition is satisfied. At least, the above example works without any iteration. Do you have an example?

Reply:
Thanks for noticing this problem. The truth is that I don't have any example that would require multiple iterations, but it seems to me that it could, in theory, occur.

However, what I realized now, is that we run this loop to make sure that the Clang parser succeeds. So I believe that we can stop collecting the incomplete types once the Clang parser runs without errors. I implemented a simple fix in [this branch](https://github.com/viktormalik/bpftrace/tree/undef-types-speedup). Could you please try it and confirm that the slowdown is gone? I'll run some tests myself and then I can open a PR.

Reply:
Yes, it drastically improves speed. 

```
% time sudo ./src/bpftrace --btf -e 'struct f { struct task_struct x; } BEGIN{exit();}'
Attaching 1 probe...


0.47s user 0.27s system 72% cpu 1.009 total
```

Please open a PR and let's discuss this further there.

Reply:
Great, thanks. 
It's not that easy after all, because this simple fix breaks one regression test. The problem is that the type resolution must be run at least once. I'll come with a better fix and open a PR.